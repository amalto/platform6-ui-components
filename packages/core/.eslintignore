dist/
loader/

src/components.d.ts