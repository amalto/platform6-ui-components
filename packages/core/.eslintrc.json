{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
