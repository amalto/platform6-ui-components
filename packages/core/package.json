{
  "name": "@platform6/components",
  "version": "0.0.1",
  "description": "Platform 6 UI components",
  "repository": {
    "type": "git",
    "url": "https://github.com/amalto/platform6-ui-components"
  },
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/custom-elements/index.js",
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "es2015": "dist/esm/index.mjs",
  "es2017": "dist/esm/index.mjs",
  "unpkg": "dist/stencil-component/stencil-component.js",
  "types": "dist/types/index.d.ts",
  "files": [
    "dist/",
    "loader/"
  ],
  "scripts": {
    "build": "stencil build --docs",
    "format": "prettier --write ./",
    "postformat": "npm run lint.fix",
    "generate": "stencil generate",
    "lint": "npm run lint.ts && npm run lint.sass",
    "lint.fix": "npm run lint.ts.fix && npm run lint.sass.fix",
    "lint.sass": "stylelint 'src/**/*.scss'",
    "lint.sass.fix": "npm run lint.sass -- --fix",
    "lint.ts": "eslint --cache --ext .js,.jsx,.ts,.tsx ./",
    "lint.ts.fix": "npm run lint.ts -- --fix",
    "start": "stencil build --docs --dev --watch --serve",
    "test": "stencil test --spec --e2e",
    "test.watch": "stencil test --spec --e2e --watchAll",
    "storybook": "start-storybook -p 6006 -s ./dist --ci",
    "build-storybook": "build-storybook"
  },
  "dependencies": {
    "@creativebulma/bulma-tooltip": "^1.2.0",
    "@fortawesome/fontawesome-svg-core": "^1.2.32",
    "@fortawesome/free-brands-svg-icons": "^5.15.1",
    "@fortawesome/free-regular-svg-icons": "^5.15.1",
    "@fortawesome/free-solid-svg-icons": "^5.15.1",
    "@stencil/core": "2.2.0",
    "bulma": "^0.9.1",
    "bulma-calendar": "^6.0.9",
    "choices.js": "9.0.1",
    "lit-html": "^1.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@stencil/eslint-plugin": "^0.3.1",
    "@stencil/postcss": "^1.0.1",
    "@stencil/react-output-target": "0.0.9",
    "@stencil/sass": "^1.3.2",
    "@storybook/addon-a11y": "^6.0.28",
    "@storybook/addon-actions": "^6.0.28",
    "@storybook/addon-essentials": "^6.0.28",
    "@storybook/addon-links": "^6.0.28",
    "@storybook/html": "^6.0.28",
    "@storybook/preset-scss": "^1.0.3",
    "@storybook/theming": "^6.0.28",
    "@storybook/web-components": "^6.0.28",
    "@types/jest": "^26.0.15",
    "@types/js-beautify": "^1.11.0",
    "@typescript-eslint/eslint-plugin": "^4.6.1",
    "@typescript-eslint/parser": "^4.6.1",
    "autoprefixer": "^9.8.6",
    "babel-loader": "^8.1.0",
    "css-loader": "^5.0.0",
    "eslint-plugin-jest": "^24.1.0",
    "jest": "^26.6.3",
    "jest-cli": "^26.6.3",
    "js-beautify": "^1.13.0",
    "prettier": "^2.1.2",
    "puppeteer": "^5.4.1",
    "react-is": "^17.0.1",
    "sass-loader": "^10.0.5",
    "style-loader": "^2.0.0",
    "typescript": "4.0.5"
  }
}
