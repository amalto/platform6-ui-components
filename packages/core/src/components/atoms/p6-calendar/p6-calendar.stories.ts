import { Components } from '../../../components';
import { ComponentProps, getElement, getForm, getSelectArgType, makeStory, Props, StringSelectArgType } from '../../../shared/storybook';
import { Mode } from '../../../shared/types';
import { P6CalendarType } from './p6-calendar';

const component = 'p6-calendar';

export default {
  title: 'Atoms/Calendar',
  component,
};

const componentProps: ComponentProps = [
  'name',
  'type',
  'required',
  'disabled',
  'color',
  'readOnly',
  'isRange',
  'startDate',
  'endDate',
  'minDate',
  'maxDate',
  'minuteSteps',
  'labelFrom',
  'labelTo',
];

const getStoryField = (props?: Props<Components.P6Calendar>): HTMLElement => {
  return getElement(component, '', props);
};

// Default
export const Default = makeStory<{
  required: boolean;
  readOnly: boolean;
  disabled: boolean;
  mode: Mode;
  type: P6CalendarType;
  lang: string;
  isRange: boolean;
  startDate: string;
  endDate: string;
  minDate: string;
  maxDate: string;
  minuteSteps: number;
  labelFrom: string;
  labelTo: string;
}>({
  componentProps,
  args: {
    required: false,
    readOnly: false,
    disabled: false,
    mode: Mode.default,
    type: 'date',
    lang: 'en',
    isRange: false,
    startDate: '',
    endDate: '',
    minDate: '',
    maxDate: '',
    minuteSteps: 5,
    labelFrom: '',
    labelTo: '',
  },
  argTypes: {
    ...getSelectArgType<StringSelectArgType>(
      'type',
      ['date', 'datetime', 'time'].map(type => ({
        key: type,
        value: type,
      })),
    ),
  },
  builder: (props): HTMLElement => getStoryField(props),
});

export const Form = makeStory<{
  required: boolean;
  readOnly: boolean;
  disabled: boolean;
  mode: Mode;
  type: P6CalendarType;
}>({
  componentProps,
  args: {
    required: false,
    readOnly: false,
    disabled: false,
    mode: Mode.default,
    type: 'date',
  },
  argTypes: {
    ...getSelectArgType<StringSelectArgType>(
      'type',
      ['date', 'datetime', 'time'].map(type => ({
        key: type,
        value: type,
      })),
    ),
  },
  builder: (props): HTMLElement => getForm(getStoryField(props)),
});
