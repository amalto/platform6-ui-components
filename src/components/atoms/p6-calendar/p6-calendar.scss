@import '~bulma/bulma';
@import '~bulma-calendar/src/sass/index';

$calendar-input-padding-top: 0.275rem;

:host {
  display: block;

  .datetimepicker-dummy {
    &::before {
      content: none;
    }

    .datetimepicker-dummy-wrapper {
      @include input;

      border-radius: $input-radius;

      .datetimepicker-dummy-input {
        font-size: $size-normal;
        padding-top: 0.275rem;
      }

      .datetimepicker-dummy-input:first-child {
        text-indent: 0.5rem;
      }
    }
  }

  .datetimepicker-clear-button {
    font-size: $size-medium;
    top: $calendar-input-padding-top * -1;
  }

  @each $name, $pair in $colors {
    $color: nth($pair, 1);
    .datetimepicker.is-#{$name} {
      .datetimepicker-footer-today {
        color: $color !important;
      }

      .datepicker-body {
        .datepicker-months .datepicker-month,
        .datepicker-years .datepicker-year {
          &.is-active {
            border-color: $color;
            background: $color;
          }

          &:hover {
            border-color: $color;
            background: #fff;
          }
        }
      }
    }
  }
}

:host([disabled]:not([disabled='false'])),
:host([readonly]:not([readonly='false'])) {
  .datetimepicker-dummy .datetimepicker-clear-button {
    display: none;
  }
}

:host([disabled]:not([disabled='false'])) {
  .datetimepicker-dummy .datetimepicker-dummy-wrapper {
    background-color: $input-disabled-background-color;
    border-color: $input-disabled-border-color;
    color: $input-disabled-color;

    &,
    .datetimepicker-dummy-input {
      cursor: not-allowed;
    }
  }
}
