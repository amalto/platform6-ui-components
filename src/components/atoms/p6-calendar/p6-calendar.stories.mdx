import { Story, Preview, Meta } from "@storybook/addon-docs/blocks";
import {
  boolean,
  select,
  number,
  text as kText,
  withKnobs,
} from "@storybook/addon-knobs";
import { action, decorate } from "@storybook/addon-actions";
import readme from "./readme.md";

<Meta
  title="Atoms/Calendar"
  component="p6-calendar"
  parameters={{ notes: readme }}
/>

# Calendar

<Story
  name="Default"
  parameters={{ docs: { disable: true } }}
  decorators={[withKnobs]}
>
  {() => {
    const kRequired = boolean("Required", false);
    const kReadOnly = boolean("Read only", false);
    const kDisabled = boolean("Disabled", true);
    const kColor = select(
      "Color",
      ["default", "primary", "info", "success", "warning", "danger"],
      "danger"
    );
    const kType = select("Type", ["date", "datetime", "time"], "date");
    const kLang = select("Language", ["en", "fr", "ru"], "en");
    const kIsRange = boolean("Is range", true);
    const kStartDate = kText("Start date", "");
    const kEndDate = kText("End date", "");
    const kMinDate = kText("Min date", "");
    const kMaxDate = kText("Max date", "");
    const kMinuteSteps = number("Minute step", 5);
    const kLabelFrom = kText("Label from", "");
    const kLabelTo = kText("Label to", "");
    return `
      <p6-calendar
        name="date"
        lang="${kLang}"
        type="${kType}"
        color="${kColor}"
        readOnly=${kReadOnly}
        required=${kRequired}
        disabled=${kDisabled}
        isRange=${kIsRange}
        startDate="${kStartDate}"
        endDate="${kEndDate}"
        minDate="${kMinDate}"
        maxDate="${kMaxDate}"
        minuteSteps=${kMinuteSteps}
        labelFrom="${kLabelFrom}"
        labelTo="${kLabelTo}"/>
    `;
  }}
</Story>
