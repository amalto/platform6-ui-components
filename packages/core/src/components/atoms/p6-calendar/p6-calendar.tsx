import { Component, ComponentInterface, Element, h, Host, JSX, Prop } from '@stencil/core';
import bulmaCalendar from 'bulma-calendar';
import { Mode } from '../../../shared/types';
import { toDate } from '../../../utils/attribute';
import { getClosestLanguage } from '../../../utils/language';
import { getL10n, L10n } from '../../../utils/translations';

export type P6CalendarType = 'date' | 'datetime' | 'time';

@Component({
  tag: 'p6-calendar',
  styleUrls: ['p6-calendar.scss'],
  shadow: true,
})
export class P6Calendar implements ComponentInterface {
  /**
   * The name
   */
  @Prop() public name!: string;

  /**
   * Type of field (date, time, datetime)
   */
  @Prop() public type: P6CalendarType = 'date';

  /**
   * Required - If `true`, the user must set a value to be valid
   */
  @Prop() required = false;

  /**
   * Disabled - If `true`, the user cannot interact with the field.
   */
  @Prop() disabled = false;

  /**
   * The color of the field
   */
  @Prop() color: Mode = Mode.info;

  /**
   * Marks the field as read only
   */
  @Prop({ attribute: 'readOnly' }) readOnly = false;

  /**
   * Range capability (start and end date/time selection)
   */
  @Prop({ attribute: 'isRange' }) isRange = false;

  /**
   * Pre-selected start date
   */
  @Prop({ attribute: 'startDate' }) startDate: string | undefined;

  /**
   * Pre-selected end date
   */
  @Prop({ attribute: 'endDate' }) endDate: string | undefined;

  /**
   * Minimum date allowed
   */
  @Prop({ attribute: 'minDate' }) minDate: string | undefined;

  /**
   * Maximum date allowed
   */
  @Prop({ attribute: 'maxDate' }) maxDate: string | undefined;

  /**
   * Steps for minutes selector
   */
  @Prop({ attribute: 'minuteSteps' }) minuteSteps = 5;

  /**
   * From label
   */
  @Prop({ attribute: 'labelFrom' }) labelFrom = '';

  /**
   * To label
   */
  @Prop({ attribute: 'labelTo' }) labelTo = '';

  @Element() private readonly host!: HTMLP6CalendarElement;

  private nativeInput: HTMLInputElement | undefined;

  private calendar: bulmaCalendar | undefined;

  private l10n: L10n | undefined;

  async componentWillLoad(): Promise<void> {
    this.l10n = await getL10n(this.host);
  }

  render(): JSX.Element {
    const attributes = {
      name: this.name,
    };

    return (
      <Host aria-disabled={this.disabled ? 'true' : null} disabled={this.disabled} readonly={this.readOnly}>
        <input
          ref={(input: HTMLInputElement | undefined): void => {
            this.nativeInput = input;
          }}
          {...attributes}
        />
      </Host>
    );
  }

  componentDidRender(): void {
    if (this.nativeInput === undefined) {
      return;
    }

    // eslint-disable-next-line new-cap
    this.calendar = new bulmaCalendar(this.nativeInput, {
      isRange: this.isRange,
      type: this.type,
      startDate: toDate(this.startDate),
      endDate: toDate(this.endDate),
      minDate: toDate(this.minDate),
      maxDate: toDate(this.maxDate),
      minuteSteps: this.minuteSteps,
      labelFrom: this.labelFrom,
      labelTo: this.labelTo,
      toggleOnInputClick: !this.readOnly && !this.disabled,
      showClearButton: false,
      color: Mode[this.color],
      dateFormat: 'YYYY-MM-DD',
      lang: getClosestLanguage(this.nativeInput),
      clearLabel: this.l10n?.clearLabel,
      todayLabel: this.l10n?.todayLabel,
      cancelLabel: this.l10n?.cancelLabel,
      nowLabel: this.l10n?.nowLabel,
      validateLabel: this.l10n?.validateLabel,
    });
  }

  disconnectedCallback(): void {
    this.calendar?.hide();
    this.calendar = undefined;
  }
}
