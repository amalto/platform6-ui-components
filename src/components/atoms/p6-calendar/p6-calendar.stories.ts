import { Components } from "../../../components";
import {
  getElement,
  getSelectArgType,
  makeStory,
  Props,
  StringSelectArgType,
} from "../../../shared/storybook/stories";
import { Mode } from "../../../shared/types";
import { P6CalendarType } from "./p6-calendar";

const component = "p6-calendar";

export default {
  title: "Atoms/Button",
  component,
};

const getStoryField = (props?: Props<Components.P6Calendar>): HTMLElement => {
  return getElement(component, "", props);
};

// Default
export const Default = makeStory<{
  required: boolean;
  readOnly: boolean;
  disabled: boolean;
  mode: Mode;
  type: P6CalendarType;
  lang: string;
  isRange: boolean;
  startDate: string;
  endDate: string;
  minDate: string;
  maxDate: string;
  minuteSteps: number;
  labelFrom: string;
  labelTo: string;
}>({
  args: {
    required: false,
    readOnly: false,
    disabled: false,
    mode: Mode.default,
    type: "date",
    lang: "en",
    isRange: false,
    startDate: "",
    endDate: "",
    minDate: "",
    maxDate: "",
    minuteSteps: 5,
    labelFrom: "",
    labelTo: "",
  },
  argTypes: {
    ...getSelectArgType<StringSelectArgType>(
      "type",
      ["date", "datetime", "time"].map((type) => ({
        key: type,
        value: type,
      }))
    ),
  },
  builder: (props): HTMLElement => getStoryField(props),
});
