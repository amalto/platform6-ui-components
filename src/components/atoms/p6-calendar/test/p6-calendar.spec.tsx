import { newSpecPage } from "@stencil/core/testing";
import { P6Calendar } from "../p6-calendar";

describe("p6-calendar", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [P6Calendar],
      html: `<p6-calendar></p6-calendar>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
