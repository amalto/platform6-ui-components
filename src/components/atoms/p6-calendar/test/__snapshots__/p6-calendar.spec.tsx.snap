// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-calendar renders 1`] = `
<p6-calendar>
  <input>
</p6-calendar>
`;

exports[`p6-calendar translation Should translate en 1`] = `
<div lang="en">
  <p6-calendar></p6-calendar>
</div>
`;

exports[`p6-calendar translation Should translate fr 1`] = `
<div lang="fr">
  <p6-calendar></p6-calendar>
</div>
`;
